embassy-nrf/src/low_power_uarte.rs

```diff
@@ -0,0 +1,302 @@
+//! Async low power UARTE.
+//!
+//! The peripheral is autmatically enabled and disabled as required to save power.
+//! Lowest power consumption can only be guaranteed if the send receive futures
+//! are dropped correctly (e.g. not using `mem::forget()`).
+
+use core::future::Future;
+use core::sync::atomic::{compiler_fence, Ordering};
+use core::task::{Context, Poll};
+
+use embassy::util::Signal;
+use embedded_dma::{StaticReadBuffer, StaticWriteBuffer};
+
+use crate::fmt::{assert, trace};
+use crate::hal::gpio::{Floating, Input, Output, Pin, PushPull};
+use crate::hal::pac;
+use crate::hal::prelude::*;
+use crate::hal::target_constants::EASY_DMA_SIZE;
+use crate::interrupt;
+
+// Re-export SVD variants to allow user to directly set values.
+pub use pac::uarte0::{baudrate::BAUDRATE_A as Baudrate, config::PARITY_A as Parity};
+
+/// Interface to the UARTE peripheral
+pub struct LowPowerUarte {
+    instance: pac::UARTE0,
+}
+
+struct State {
+    tx_done: Signal<()>,
+    rx_done: Signal<u32>,
+}
+
+static STATE: State = State {
+    tx_done: Signal::new(),
+    rx_done: Signal::new(),
+};
+
+impl LowPowerUarte {
+    pub fn new(uarte: pac::UARTE0) -> Self {
+        // Enable interrupts
+        uarte.events_endtx.reset();
+        uarte.events_endrx.reset();
+        uarte
+            .intenset
+            .write(|w| w.endtx().set().txstopped().set().endrx().set().rxto().set());
+        // TODO: Set interrupt priority?
+        interrupt::unpend(interrupt::UARTE0_UART0);
+        interrupt::enable(interrupt::UARTE0_UART0);
+
+        LowPowerUarte { instance: uarte }
+    }
+
+    /// Sets the baudrate, parity and assigns the pins to the UARTE peripheral.
+    // TODO: Make it take the same `Pins` structs nrf-hal (with optional RTS/CTS).
+    // TODO: #[cfg()] for smaller device variants without port register (nrf52810, ...).
+    pub fn configure(
```

Comment:
if is a separate method, it's possible to create an instance and use it directly without configuring first. `BufferedUarte` takes all the config in `new` for this reason.

It can be useful to have a `configure` method to change the config, but `new` must leave the instance in a ready-to-use state

Reply:
nrf-hal and `BufferedUarte` take ownership of the pins.
I opted for this solution (pins by ref) because I need to time multiplex the UART peripheral between multiple pins.

Looking at the datasheet I see that the peripheral can indeed operate without an explicit call to `configure()`.
It then uses the default values of 250000bps and no parity.

Anyway, I can just mirror the parameters to the `new()` function and call `configure()` from there if you prefer that :)

Reply:
It seems to me very strange to rely on the defaults. And I think for the pins it'll have them all disconnected. It doesn't make sense to allow the user to use the uart in that broken state.